go/grpcweb/wrapper.go

```diff
@@ -193,11 +205,61 @@ func (w *WrappedGrpcServer) isRequestForRegisteredEndpoint(req *http.Request) bo
 	return false
 }
 
-func hackIntoNormalGrpcRequest(req *http.Request) *http.Request {
+// Returns a reader that will read base64 encoded bytes from r, decode them, and hand them back.
+// This starts a goroutine to avoid excessive buffering.
+func decodeBase64Reader(r io.ReadCloser) io.ReadCloser {
+	pipeReader, pipeWriter := io.Pipe()
+	decoder := base64.NewDecoder(base64.StdEncoding, r)
+
+	go func() {
+		// read from decoder, put data into the pipe; always close the original reader
+		_, err := io.Copy(pipeWriter, decoder)
+		closeErr := r.Close()
+		if err != nil {
+			// error occurred: return it on the reader side
+			pipeReader.CloseWithError(err)
+		}
+		if closeErr != nil {
+			pipeReader.CloseWithError(closeErr)
+		}
+
+		// decoded everything: close the pipe
+		err = pipeWriter.Close()
+		if err != nil {
+			pipeReader.CloseWithError(err)
+		}
+	}()
+
+	return pipeReader
+}
+
+type readerCloser struct {
+	reader io.Reader
+	closer io.Closer
+}
+
+func (r *readerCloser) Read(dest []byte) (int, error) {
+	return r.reader.Read(dest)
+}
+func (r *readerCloser) Close() error {
+	return r.closer.Close()
+}
+
+func hackIntoNormalGrpcRequest(req *http.Request) (*http.Request, bool) {
 	// Hack, this should be a shallow copy, but let's see if this works
 	req.ProtoMajor = 2
 	req.ProtoMinor = 0
+
 	contentType := req.Header.Get("content-type")
-	req.Header.Set("content-type", strings.Replace(contentType, "application/grpc-web", "application/grpc", 1))
-	return req
+	incomingContentType := grpcWebContentType
+	isTextFormat := strings.HasPrefix(contentType, grpcWebTextContentType)
+	if isTextFormat {
+		// body is base64-encoded: decode it; Wrap it in readerCloser so Body is still closed
+		decoder := base64.NewDecoder(base64.StdEncoding, req.Body)
+		req.Body = &readerCloser{decoder, req.Body}
```

Comment:
Nit: can we make this explicitly assign the fields for clarity, I think this is a vet error now?

Reply:
Done! Although this doesn't trigger any vet warnings, since this is a struct declared in this package. Declaring too many/few fields causes a compiler error with "too few values in &readerCloser literal", or if this was in a different package vet would warn about "composite literal uses unkeyed fields".

I also took this opportunity to tweak the initialization of `grpcWebResponse` in `newGrpcWebResponse` and remove an unused function from my previous io.Pipe related misadventure.